### PR TITLE
[SHIMGVW] Add 'rotate and save image' buttons

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Improve compatibility.
JIRA issue: [CORE-19358](https://jira.reactos.org/browse/CORE-19358), [CORE-19387](https://jira.reactos.org/browse/CORE-19387)

## Proposed changes

- Add `IDC_ROT_CWSAVE` and `IDC_ROT_CCWSAVE` buttons.
- Add `Preview_pSaveImage` helper function.
- Use it in `IDC_ROT_CWSAVE` and `IDC_ROT_CCWSAVE` command handling.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Do we support **lossless** rotation yet, and keeping all the existing metadata upon saving? Otherwise I would propose to implement that first, because XPSP3/2k3sp2 stupid default behavior of shreddering the data being viewed upon rotation is what personally never ever press that button in the older versions of Windows.

If we wouldn't support both yet, I would vote to have a 2nd button for "rotate and save" in the meantime as a temporary solution.
We shouldn't follow the really bad example of the older Windows versions here of being a data-destructor by calling a operation that seems to be innocent to the users eye while in fact it is not.

**Comment by JoachimHenze** [comment]
And a question: What happens when rotating an image on a read-only-media. Does it generate a nagging warning then or does it fail silently to save?

**Comment by katahiromz** [comment]
> we support lossless rotation

Nope.

> keeping all the existing metadata upon saving

Nope.

> I would propose to implement that first

It is difficult for me.

> What happens when rotating an image on a read-only-media

Like this:
read-only

**Comment by JoachimHenze** [comment]
> Nope. Nope. It's difficult for me.

Then let's duplicate the rotation buttons meanwhile. One that rotates and saves, and one that does just rotate but does not touch the file on the disk.

**Review by JoachimHenze — CHANGES_REQUESTED** [review]
Then let's duplicate the rotation buttons meanwhile. One that rotates and saves, and one that does just rotate in-memory but does not save & shredder the files.

**Comment by katahiromz** [comment]
無題
Duplicated buttons.

**Comment by katahiromz** [comment]
image
Duplicated buttons. Added text `"(Lossy)"`.

**Comment by katahiromz** [comment]
image
Added red dot to the icons.

**Comment by Matriks404** [comment]
Maybe add dialog box with info that lossless rotation is not implemented yet, and checkbox to not display again that dialog box for this picture again? Or is that too annoying if user wants to rotate a lot of pictures?

Of course I understand that it would be a bit harder to implement than adding few buttons :)

**Comment by HBelusca** [comment]
General question: why would the rotate (and save) functionality be lossy?

File: dll/win32/shimgvw/shimgvw.c [hunk 1]
```diff
@@ -387,7 +392,7 @@ Preview_pLoadImage(PPREVIEW_DATA pData, LPCWSTR szOpenFileName)
     /* check file presence */
     if (!szOpenFileName || GetFileAttributesW(szOpenFileName) == 0xFFFFFFFF)
     {
-        DPRINT1("File %s not found!\n", szOpenFileName);
+        DPRINT1("File %ls not found!\n", szOpenFileName);
```
Comment (HBelusca) [inline dll/win32/shimgvw/shimgvw.c:395]: or `File %S not found!\n`

File: dll/win32/shimgvw/shimgvw.c [hunk 2]
```diff
@@ -1420,17 +1475,35 @@ Preview_OnCommand(HWND hwnd, UINT nCommandID)
...
+            if (g_pImage && g_pCurrentFile)
             {
                 GdipImageRotateFlip(g_pImage, Rotate270FlipNone);
                 Preview_UpdateImage(pData);
             }
             break;
 
         case IDC_ROT_COUNCW:
-            if (g_pImage)
+            if (g_pImage && g_pCurrentFile)
```
Comment (HBelusca) [inline dll/win32/shimgvw/shimgvw.c:1486]: Why does this case (and the one _above_) needs to check for the `g_pCurrentFile` ?
(Note that I however understand why the two new other cases that are below need to check the pointer, though.)

File: dll/win32/shimgvw/shimgvw.c [hunk 3]
```diff
@@ -500,13 +555,13 @@ Preview_pSaveImageAs(PPREVIEW_DATA pData)
...
 
-        if (IsEqualGUID(&rawFormat, &codecInfo[j].FormatID) != FALSE)
+        if (IsEqualGUID(&rawFormat, &codecInfo[j].FormatID))
         {
             sfn.nFilterIndex = j + 1;
         }
     }
 
-    if (GetSaveFileNameW(&sfn))
+    if (GetSaveFileNameW(&sfn) && sfn.nFilterIndex > 0)
```
Comment (HBelusca) [inline dll/win32/shimgvw/shimgvw.c:564]:
```suggestion
    if (GetSaveFileNameW(&sfn) && (sfn.nFilterIndex > 0))
```

**Review by HBelusca — APPROVED** [review]
OK, though something needs to be clarified.

**Comment by katahiromz** [comment]
Some image file has EXIF information, multiple pages, and/or animation. And JPEG is basically lossy encryption.
GDI+ cannot easily save full information.

File: dll/win32/shimgvw/shimgvw.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shimgvw/shimgvw.c:395]:
```suggestion
        DPRINT1("File %S not found!\n", szOpenFileName);
```

File: dll/win32/shimgvw/shimgvw.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shimgvw/shimgvw.c]: Fixed by using `pData->m_szFile`.

**Pull request merged**